### ISSUE TITLE
`join()` comes before `contain()`, results in "Column not found"

### ISSUE THREAD

**Issue by mehov** [cakephp/cakephp#17680]
### Description

What I have
--

    return $this
        ->find('all')
        ->contain([
            'Categories' => [
                'Stores' => [
                ]
            ],
            'ProductsProducts' => [
                'RefProducts' => [
                ],
            ],
        ])
        ->join([
            [
                'alias' => 'RefCategories',
                'table' => 'Categories',
                'type' => 'LEFT',
                'conditions' => 'RefProducts.category_id = RefCategories.id'
            ],
        ])
    ;

What happens
--

In the query that CakePHP generates for that find, my manual join I defined last actually comes before the "contained" tables. 

```sql
SELECT 
  *
FROM 
  Products Products 
  LEFT JOIN Categories RefCategories ON RefProducts.category_id = RefCategories.id 
  LEFT JOIN Categories Categories ON Categories.id = Products.category_id 
  INNER JOIN Stores Stores ON Stores.id = Categories.store_id 
  LEFT JOIN Products_Products ProductsProducts ON ProductsProducts.id = Products.id 
  INNER JOIN Products RefProducts ON RefProducts.id = ProductsProducts.ref_product_id 
```

As a result,

> SQLSTATE[42S22]: Column not found: 1054 Unknown column 'RefProducts.category_id' in 'on clause'

What I expected to happen
--

Because I defined the `join()` after `contain()`, I expected the contained tables to be available *before* the join kicks in.

If I copy the SQL from DebugKit and manually edit it to where the contained tables are joined first, the query works when I run it in Workbench

### CakePHP Version

4.4.14

### PHP Version

8.3.1

**Comment by markstory** [comment]
This is likely caused by join() clauses being separate from contain() clauses. The reason for this separation is that not all contain() operations generate joins. #17184 is somewhat related.

**Comment by mehov** [comment]
@markstory but shouldn't it generate a bunch of left joins because it's a chain of `belongsTo` associations?

One can normally get Product to Category to Store all in one query

**Comment by markstory** [comment]
> but shouldn't it generate a bunch of left joins because it's a chain of belongsTo associations?

Sure, in your situation that is true, but that isn't always the case as contain() can be used for all association types.

**Comment by dereuromark** [comment]
Workarounds exist:
    - Use innerJoinWith() or leftJoinWith() instead of contain() for associations that need to be referenced in manual joins
    - Use matching() for filtering instead of manual joins
    - Build queries in stages to control join order
    - Use subqueries instead of joins where possible

We should probably document this though.

 The best practice is to use innerJoinWith() or leftJoinWith() when you need to reference an association in a manual join, as these methods add their joins immediately rather than deferring them

@markstory What do you think?

**Comment by markstory** [comment]
> What do you think?

This is another issue where, I can agree that the current behavior isn't ideal. I'm unlikely to fix this as it doesn't impact me, and from the lack of activity it doesn't impact others enough for a solution to be found.

**Comment by ADmad** [comment]
I don't think this is a solvable issue. If someone needs specific ordering of joins then they should use one of the joining methods for all joins instead of using `contain()`.

**Comment by dereuromark** [comment]
Closing then as workarounds exist.
